rust-symcrypt/src/rsa/mod.rs

```diff
@@ -0,0 +1,716 @@
+/// NOTE: SymCrypt does not pre-append leading 0's if the MSB is set.
+pub struct RsaPublicKeyExportBlob {
+    pub modulus: Vec<u8>,
+    pub pub_exp: Vec<u8>,
+}
+
+/// Impl for RsaKeyPair struct.
+impl RsaKeyPair {
+    /// `generate_new()` generates a random Rsa key based on the provided parameters.
+    ///
+    /// `n_bits_mod` represents a `u32` that is the desired bit length of the Rsa key, `n_bits_mod` must be at least 1024 bits.
+    ///
+    ///
+    /// `pub_exp` takes in an `Option<u64>` that is the public exponent. If `None` is provided, the default `2^16 +1` will be used.
+    ///  
+    /// `key_usage` takes in a [`RsaKeyUsage`] and will indicate if this key will be used for [`RsaKeyUsage::Sign`], or [`RsaKeyUsage::Encrypt`], or [`RsaKeyUsage::SignAndEncrypt`]
+    pub fn generate_new(
+        n_bits_mod: u32,
+        pub_exp: Option<&[u8]>,
+        key_usage: RsaKeyUsage,
+    ) -> Result<Self, SymCryptError> {
+        let (pub_exp_ptr, pub_exp_count) = match pub_exp {
+            Some(exp) => {
+                let u64_pub_exp = load_msb_first_u64(exp)?;
+                ([u64_pub_exp].as_ptr(), 1) // This array has a length of 1.
+            }
+            None => (ptr::null(), 0), // If no public exponent is provided, use null and count 0 which will notify SymCrypt to use their default exponent.
+        };
+
+        let rsa_key = allocate_rsa(2, n_bits_mod)?;
+
+        unsafe {
+            // SAFETY: FFI calls
+            match symcrypt_sys::SymCryptRsakeyGenerate(
+                rsa_key.0,
+                pub_exp_ptr,   // Pointer to the public exponent array or null.
+                pub_exp_count, // Count of public exponents. Will be 1 if Some() is provided, else 0.
+                key_usage.to_symcrypt_flag(),
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(RsaKeyPair {
+                    inner: rsa_key,
+                    key_usage: key_usage,
+                }),
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    /// `set_key_pair()` sets both the public and private key information onto [`RsaKeyPair`].
+    ///
+    /// `modulus_buffer` takes in a `&[u8]` reference to a byte array that contains the modulus of the Rsa key.
+    ///
+    /// `pub_exp` takes in a `&[u8]` that is the public exponent represented by an array of bytes.
+    ///
+    /// `p` takes in a `&[u8]` reference to a byte array that contains the first prime.
+    ///
+    /// `q` takes in a `&[u8]` reference to a byte array that contains the second prime.
+    pub fn set_key_pair(
+        modulus_buffer: &[u8],
+        pub_exp: &[u8],
+        p: &[u8],
+        q: &[u8],
+        key_usage: RsaKeyUsage,
+    ) -> Result<Self, SymCryptError> {
+        let n_bits_mod = (modulus_buffer.len() as u32) * 8; // Convert the size from bytes to bits.
+        let rsa_key = allocate_rsa(2, n_bits_mod)?;
+        let u64_pub_exp = load_msb_first_u64(pub_exp)?;
```

Comment:
If you require the caller to pass in a pub_exp that is u64 value, then, you will need to provide a public function for and document.